lib/Agrammon/OutputFormatter/CSV.pm6

```diff
@@ -1,51 +1,94 @@
+                        add-filters($raw-value, $module, $var, $unit, $prefix, :$all-filters);
+                    }
+                }
             }
-         }
-        else {
-            for .value.sort(*.key) {
-                my $instance-id = .key;
-                for .value.sort(*.key) {
-                    my $fq-name = .key;
+        }
+        when Array {
+            for sorted-kv($_) -> $instance-id, %instance-outputs {
+                my $filters = '';
+                for sorted-kv(%instance-outputs) -> $fq-name, %values {
                     my $q-name = $module ~ '[' ~ $instance-id ~ ']' ~ $fq-name.substr($module.chars);
-                    for .value.sort(*.key) {
+                    for sorted-kv(%values) -> $output, $value {
+                        next unless $model.should-print($module, $output, @print-set);
+
+                        my $raw-value = $value;
+                        my $unit = $model.output-unit($module, $output, $language);
                         my @data = (
-                            $q-name, .key, flat-value(.value) // '',
-                            $model.output-unit($fq-name, .key, $unit-language)
+                            |($prefix if $prefix),
+                            $q-name,
+                            $output,
+                            $filters,
+                            flat-value($raw-value) // '',
+                            $unit
                         );
-                        @data.unshift($prefix) if $prefix;
                         take @data.join(';');
+                        if $include-filters {
+                            if $raw-value ~~ Agrammon::Outputs::FilterGroupCollection && $raw-value.has-filters {
+                                add-filters($raw-value, $q-name, $fq-name, $unit, $prefix, :$all-filters);
+                            }
+                        }
                     }
                 }
             }
         }
-    }).join("\n") ~ "\n";
+    }).join("\n");
 }
 
-multi sub flat-value($value) {
-    $value
-}
-multi sub flat-value(Agrammon::Outputs::FilterGroupCollection $collection) {
-    +$collection
+sub add-filters(Agrammon::Outputs::FilterGroupCollection $collection,
+        $q-name, $fq-name, $unit, Str $prefix, Bool :$all-filters) {
+    my @results = $collection.results-by-filter-group(:all($all-filters));
+    for @results {
+        my %filters := .key;
+        my $raw-value := .value;
+        my $filters = %filters.values.join(',');
+        my @filters = %filters.map: { .key ~ '=' ~ .value };
+        for (@filters || '(Uncategorized)') {
+            my @data = (
+                |($prefix if $prefix),
+                $q-name,
+                $fq-name,
+                $filters,
+                flat-value($raw-value) // '',
+                $unit
+            );
+            take @data.join(';');
+        }
```

Comment:
I'm still a bit confused about this; `$_` is not used in the loop either, so it will just output the very same data every iteration of the loop?

Reply:
Hmm, it does the right thing, because `@filters` only has one element (we don't use multiple filter groups right now).
But you're quite right, I think I fixed it appropriately for now.

It will need some change if we ever use multiple filter groups, so I added a warning about that.